openedx/core/djangoapps/appsembler/sites/tasks.py

```diff
@@ -1,58 +1,106 @@
 LOGGER = get_task_logger(__name__)
 FULL_COURSE_REINDEX_THRESHOLD = 1
 
 
 @task()
-def clone_course(source_course_key_string, destination_course_key_string, user_id, fields=None):
+def import_course_on_site_creation(organization):
     """
     Reruns a course in a new celery task.
     """
-    # import here, at top level this import prevents the celery workers from starting up correctly
-    from edxval.api import copy_course_videos
-    from contentstore.utils import initialize_permissions
-    from contentstore.courseware_index import SearchIndexingError
-    from contentstore.views.course import reindex_course_and_check_access
-
     try:
-        # deserialize the payload
-        source_course_key = CourseKey.from_string(source_course_key_string)
-        destination_course_key = CourseKey.from_string(destination_course_key_string)
-
-        # use the split modulestore as the store for the rerun course,
-        # as the Mongo modulestore doesn't support multiple runs of the same course.
-        store = modulestore()
-        with store.default_store('split'):
-            store.clone_course(source_course_key, destination_course_key, user_id, fields=fields)
-
-        # Send statistics
-        from appsembler.models import update_course_statistics
-        update_course_statistics()
-
-        # set initial permissions for the user to access the course.
-        user = User.objects.get(id=user_id)
-        initialize_permissions(destination_course_key, user)
-
-        # add course intructor and staff roles to the new user
-        CourseInstructorRole(destination_course_key).add_users(user)
-        CourseStaffRole(destination_course_key).add_users(user)
-
-        # call edxval to attach videos to the rerun
-        copy_course_videos(source_course_key, destination_course_key)
+        course_name = settings.TAHOE_DEFAULT_COURSE_NAME
+        course_github_org = settings.TAHOE_DEFAULT_COURSE_GITHUB_ORG
+        course_github_name = settings.TAHOE_DEFAULT_COURSE_GITHUB_NAME
+        course_version = settings.TAHOE_DEFAULT_COURSE_VERSION
+
+        # create a dir in tmp to download the default course
+        dir = os.path.join("/", "tmp", str(uuid.uuid4()))
+        if not os.path.exists(dir):
+            os.mkdir(dir)
+
+        # build the github download URL from settings to download the course
+        # in tar.gz format from the github releases
+        course_download_url = 'https://github.com/{}/{}/archive/{}.tar.gz'.format(
+            course_github_org,
+            course_github_name,
+            course_version
+        )
+
+        # download the course in tar.gz format from github in /tmp using a
+        # random file name
+        r = requests.get(course_download_url)
+        course_file_name = str(uuid.uuid4()) + ".zip"
+        file_path = os.path.join("/", "tmp", course_file_name)
+        with open(file_path, 'wb') as fd:
+            for chunk in r.iter_content(chunk_size=128):
+                fd.write(chunk)
```

Comment:
Please consider the [`copyfileobj` shortcut](https://stackoverflow.com/a/39217788/161278). I really hate for-loops for data transfer, it makes me think that I'm writing a C program:

```suggestion
import requests
import shutil

def download_file(url):
    local_filename = url.split('/')[-1]
    with requests.get(url, stream=True) as r:
        with open(local_filename, 'wb') as f:
            shutil.copyfileobj(r.raw, f)

    return local_filename
```

Reply:
@OmarIthawi It looks like C because it's old, we're using an old version of `requests` in edx that doesn't allow streams, so I tried it but it was failing, please check this: https://github.com/psf/requests/issues/4157 it was going to be released in 2017, and we're using a version from 2015.